chore(spanner): Add benchwrapper.

rajatbhatta opened the pull request:
This PR adds a Benchwrapper implementation for Python Spanner Client Library.

The benchwrapper implementation contains the following:

- The proto file, as well as the generated classes from the proto file.
- A file that connects to the emulator using the SPANNER_EMULATOR_HOST environment variable, and implements the RPC’s defined in the proto file.

Checks:

- [x] Appropriate docs were added, wherever necessary.
- [x] Ensure the tests and linter pass.

rajatbhatta committed:
chore(spanner): Add benchwrapper.

rajatbhatta committed:
Add README.md to run the benchwrapper

rajatbhatta committed:
Add README for regenerating protos for benchmarking

rajatbhatta committed:
Small change to benchwrapper README

rajatbhatta committed:
Merge branch 'main' into spanner-client-libs-benchmarking

zoercai commented on benchmark/benchwrapper/README.md:
````diff
@@ -0,0 +1,10 @@
+## Running
+Run the following commands from python-spanner/ directory.
+```
+export SPANNER_EMULATOR_HOST=localhost:8080
````
Is `8080` the port you used while testing? The other client libraries are using `9010`

rajatbhatta replied:
Thanks for pointing it out, Zoe! I've updated this to 9010, we can have it consistent across client libraries.

zoercai reviewed (commented):
Please also get an approval from a Python reviewer before merging, my LGTM is only for what's been implemented, not the Python syntax/readability.

rajatbhatta committed:
Update README.md for benchwrapper.

IlyaFaer commented on benchmark/benchwrapper/main.py:
```diff
@@ -0,0 +1,157 @@
+# Copyright 2022 Google, Inc.
```
There are several different copyrights in this PR. Some are for Google Inc 2022, some for Google LLC 2019 and so on. Should be checked for which is right.

vi3k6i5 replied:
I think it's best if we stick to 1 format for the entire projec. **Do not modify any existing files**.
Just check which is the common format in this project and across projects (search on VS code/ Github) for copyright word. Use the format which is commonly used.

rajatbhatta replied:
I'll change this to "Copyright 2022 Google LLC" to make it consistent with the other files in the repository.

Not changing the copyright in spanner.proto file as it has been copied over from another internal repository.

rajatbhatta replied:
Done.

IlyaFaer commented on benchmark/benchwrapper/main.py:
```diff
@@ -0,0 +1,157 @@
+
+    validate_opts(opts)
+
+    if "SPANNER_EMULATOR_HOST" not in os.environ:
+        raise ValueError("This benchmarking server only works when connected "
+                         "to an emulator. Please set SPANNER_EMULATOR_HOST.")
```
I suppose this check should be the first in the function. `get_opts()` and `validate_opts()` are fast enough, but they are useless without the env var set.

rajatbhatta replied:
Makes sense, done.

rajatbhatta commented on benchmark/benchwrapper/main.py:
```diff
@@ -0,0 +1,157 @@
+            # do nothing with the data
+            continue
+
+        return spanner_messages.EmptyResponse()
```
The intention of the benchwrapper is to compute the time taken for the read query to be executed, and not the actual results.

The protos written are common for every language, and mandates returning of an empty response message once the function has completed execution.

rajatbhatta replied:
Done.

rajatbhatta commented on benchmark/benchwrapper/main.py:
```diff
@@ -0,0 +1,157 @@
+
+        return spanner_messages.EmptyResponse()
+
+    def Update(self, request, context):
```
Does having something like below work? Even though the "context" parameter is not used, it is still passed in while making the GRPC call. So, removing the unused parameter will cause issues.

def Update(self, request, _):

Adding @vi3k6i5 to comment on the proposal as well.

rajatbhatta replied:
Replaced context by _ as we do not care about that parameter.

rajatbhatta committed:
chore(spanner): Incorporate review comments.

rajatbhatta commented on benchmark/benchwrapper/main.py:
```diff
@@ -0,0 +1,157 @@
+        Node's Transaction.Read.
+
+        It will typically be used to read many items.
+        """
```
Done.

rajatbhatta commented on benchmark/benchwrapper/README.md:
```diff
@@ -0,0 +1,10 @@
+# Benchwrapper
+
+A small gRPC wrapper around the spanner client library. This allows the
+benchmarking code to prod at spanner without speaking Python.
```
Done.

rajatbhatta commented on benchmark/benchwrapper/main.py:
```diff
@@ -0,0 +1,157 @@
+# limitations under the License.
+
+"""This script wraps the client library in a gRPC interface that a benchmarker
+can communicate through."""
```
What change do we expect in the module docstring here?

vi3k6i5 replied:
I think this docstring should not be there, we write doc inside class and methods.

rajatbhatta replied:
As discussed offline, I'll add one line documentation along with script usage instructions as per https://engdoc.corp.google.com/eng/doc/devguide/py/style/index.md?cl=head#s3.8.1-comments-in-doc-strings.

rajatbhatta replied:
Done.

rajatbhatta commented on benchmark/benchwrapper/main.py:
```diff
@@ -0,0 +1,157 @@
+
+        for _ in results:
+            # do nothing with the data
+            continue
```
Done.

rajatbhatta commented on benchmark/benchwrapper/main.py:
```diff
@@ -0,0 +1,157 @@
+        """Method to execute batch_update in a transaction."""
+        status, row_cts = transaction.batch_update(stmts)
+
+        print(status, row_cts)
```
Removed print.

IlyaFaer commented on benchmark/benchwrapper/main.py:
```diff
@@ -38,41 +38,62 @@
+    methods.
+
+    :type project_id: str
+    :param project_id: Spanner project
```
Period in the end

rajatbhatta replied:
Done.

IlyaFaer reviewed (commented):
LGTM

rajatbhatta committed:
chore(spanner): Incorporate review comments.

vi3k6i5 commented on benchmark/benchwrapper/proto/spanner.proto:
```diff
@@ -0,0 +1,73 @@
+// Copyright 2019 Google LLC
```
nit: update year

rajatbhatta replied:
Done.

vi3k6i5 commented on benchmark/benchwrapper/proto/spanner_pb2.py:
```diff
@@ -0,0 +1,367 @@
+# -*- coding: utf-8 -*-
```
nit: add copyright msg

vi3k6i5 reviewed (approved):
LGTM

rajatbhatta committed:
chore(spanner): accommodate review comments.

-   Add script docstring with usage instructions.

-   Modify copyright in spanner.proto.

ansh0l commented:
`This branch is out-of-date with the base branch` - please update before merging

rajatbhatta committed:
Merge branch 'main' into spanner-client-libs-benchmarking

rajatbhatta committed:
Merge branch 'main' into spanner-client-libs-benchmarking

rajatbhatta commented:
> `This branch is out-of-date with the base branch` - please update before merging

@ansh0l : This is done. Please merge it. Thank you!

ansh0l reviewed (approved):
LGTM.

Minor: Please check if we can name these files (`_pb2`) better

`benchmark/benchwrapper/proto/spanner_pb2.py`
`benchmark/benchwrapper/proto/spanner_pb2_grpc.py`

rajatbhatta commented:
> LGTM.
> 
> Minor: Please check if we can name these files (`_pb2`) better
> 
> `benchmark/benchwrapper/proto/spanner_pb2.py` `benchmark/benchwrapper/proto/spanner_pb2_grpc.py`

@ansh0l : These were not written manually but were generated using grpc tool, and the so the files should not be renamed, or changed in any way.

Thank you for merging the PR!
